### PR TITLE
Cancel payment action

### PR DESCRIPTION
**Pull request by ropczan**

### CODE REVIEW

File: src/Action/CancelAction.php [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+use Payum\Core\Request\Cancel;
+use Psr\Log\LoggerInterface;
+use Webmozart\Assert\Assert;
+
+class CancelAction implements ActionInterface
+{
+    private const FINAL_STATUSES = [
+        OrderStatus::Completed,
+        OrderStatus::Canceled,
+    ];
```
Comment (lukasz-falda) [inline src/Action/CancelAction.php:23]: move it to OrderStatus class,
or create a public method like `OrderStatus->isFinal()`

File: src/Action/CancelAction.php [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+     * @param Cancel $request
+     */
+    public function execute($request): void
+    {
+        RequestNotSupportedException::assertSupports($this, $request);
+
+        $model = Model::ensureArrayObject($request->getModel());
+
+        $payment = PaymentHelper::ensurePayment($request->getFirstModel());
+        Assert::notNull($payment, 'Payment must be set on cancel action.');
```
Comment (lukasz-falda) [inline src/Action/CancelAction.php:41]: `ensurePayment` method returns PaymentInterface, not null. PHPStan does not report this to you? Maybe we need to update PHPStan 🤔

File: src/Action/CancelAction.php [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        $orderId = PaymentHelper::getOrderId($model, $payment);
+        Assert::notEmpty($orderId, 'OrderId must be set on cancel action.');
+
+        try {
+            if (!$this->canCancelPayment($model, $payment)) {
+                return;
```
Comment (lukasz-falda) [inline src/Action/CancelAction.php:47]: if we can't cancel Payment we nned to know about it. Throw custom exception here like `CannotCancelPaymentException`

File: src/Action/CancelAction.php [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+            }
+
+            $this->orderRequestService->cancel($model->orderId(), PaymentHelper::getConfigKey($model, $payment));
+        } catch (\Throwable $throwable) {
+            $this->logger->critical('Cannot cancel payment.', ['exception' => $throwable]);
```
Comment (lukasz-falda) [inline src/Action/CancelAction.php:52]: You can't catch whole exceptions and only log it, we need to know it something happened. Throw custom exception `CancelPaymentException` and pass parent to it

File: src/Action/CancelAction.php [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            && $request->getModel() instanceof \ArrayAccess
+            && $request->getFirstModel() instanceof PaymentInterface;
+    }
+
+    private function canCancelPayment(Model $model, PaymentInterface $payment): bool
+    {
+        $response = $this->orderRequestService->retrieve($model->orderId(), PaymentHelper::getConfigKey($model, $payment));
+        $status = $response->orders[0]->status;
+
+        return !in_array($status, self::FINAL_STATUSES, true);
```
Comment (lukasz-falda) [inline src/Action/CancelAction.php:69]: create `OrderStatus` enum from `$status` and use method `->isFinal()` as I mention above

Comment (lukasz-falda) [reply src/Action/CancelAction.php:69]: I have some doubts about checking this on CancelAction method. Maybe we need to check it before run CancelAction but this action should proceed only cancel request 🤔

File: tests/Integration/Action/CancelActionTest.php [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+                FileTestUtil::decodeJsonFromFile(
+                    __DIR__ . '/../Request/data/retrieveOrderResponse.json'
+                )
+            )
+        );
+
+        $payment = new Payment();
+        $payment->setDetails(FileTestUtil::decodeJsonFromFile(__DIR__ . '/data/detailsWithOrderId.json'));
+        $expectedBaseDetails = $this->getExpectedBaseDetails();
+        self::assertSame($expectedBaseDetails, $payment->getDetails());
```
Comment (lukasz-falda) [inline tests/Integration/Action/CancelActionTest.php:40]: you just set the details. We need to check it again?

File: tests/Integration/Action/CancelActionTest.php [hunk 7]
```diff
@@ -0,0 +1,100 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Answear\Payum\PayU\Tests\Integration\Action;
```
Comment (lukasz-falda) [inline tests/Integration/Action/CancelActionTest.php:5]: this test seems like the unit test, not integration

File: tests/Integration/Request/CancelOrderTest.php [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+    public function createUnauthorizedTest(): void
+    {
+        $this->mockGuzzleResponse(
+            new Response(401, [], FileTestUtil::getFileContents(__DIR__ . '/data/orderUnauthorizedResponse.json'))
+        );
+
+        $this->expectException(PayUAuthorizationException::class);
+        $this->expectExceptionCode(401);
+        $this->getOrderRequestService()->cancel('WZHF5FFDRJ140731GUEST000P01', null);
+    }
```
Comment (lukasz-falda) [inline tests/Integration/Request/CancelOrderTest.php:52]: test with 400-500 code from PayU is missing

**Review by lukasz-falda — CHANGES_REQUESTED** [review]
Update the README file

File: src/Action/CancelAction.php (hunk 2 above)
Comment (ropczan) [reply src/Action/CancelAction.php]: no, PHPStan didn't report anything :(

File: src/Action/CancelAction.php (hunk 5 above)
Comment (ropczan) [reply src/Action/CancelAction.php]: Api always returns a status code of 200, regardless of whether the payment was successfully canceled or not. For example, when you send a transaction cancellation request for a transaction already paid, you get a status of 200, but after checking the status it is still paid. For this reason, I would prefer to check the status before sending the cancellation and throw a custom exception.

File: tests/Integration/Action/CancelActionTest.php (hunk 6 above)
Comment (ropczan) [reply tests/Integration/Action/CancelActionTest.php]: it is not needed, removed

File: tests/Integration/Request/CancelOrderTest.php (hunk 8 above)
Comment (ropczan) [reply tests/Integration/Request/CancelOrderTest.php:54]: added missing tests

File: src/Action/CancelAction.php (hunk 4 above)
Comment (ropczan) [reply src/Action/CancelAction.php]: I removed the `try catch` completely, it is not needed here

File: src/Action/CancelAction.php (hunk 3 above)
Comment (ropczan) [reply src/Action/CancelAction.php]: done

File: src/Enum/OrderStatus.php [hunk 9]
```diff
@@ -11,4 +11,17 @@ enum OrderStatus: string
...
+    public static function finalStatuses(): array
+    {
+        return [
+            self::Completed,
+            self::Canceled,
+        ];
+    }
+
+    public static function isFinal(OrderStatus $status): bool
+    {
```
Comment (lukasz-falda) [inline src/Enum/OrderStatus.php:24]: maybe not static will be better? #whatever

**Review by lukasz-falda — APPROVED** [review]

File: src/Enum/OrderStatus.php (hunk 9 above)
Comment (ropczan) [reply src/Enum/OrderStatus.php:24]: done :)

**Review by Wiktor6 — APPROVED** [review]

**Pull request merged**